### PR TITLE
Create keys.rst

### PR DESCRIPTION
**Pull request by hodgesto**
Added information regarding the behavior of OpenStack SSH keys.

### CODE REVIEW

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by ramereth** [comment]
[Write docs for key-based access covering various use cases](https://trello.com/c/Z3c9mVHZ/374-write-docs-for-key-based-access-covering-various-use-cases)

File: source/openpower/keys.rst [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+When we create our OpenPOWER OpenStack instance, we set up a key pair to create the machine with.
+
+.. note::
+
+  For the purposes of this documentation, this key pair will be known as the **Original Key Pair**.
+
+To allow for multiple SSH Keys to access this instance:
+  We need to add the contents of ``<our_key>.pub`` to ``~/.ssh/authorized_keys`` from the machine with the corresponding keyfile.
+  
+  To do so, we need to be certain we can ``ssh`` into our instance with::
```
Comment (doublej472) [inline source/openpower/keys.rst]: You can add your key info way easier with the `ssh-copy-id` command.

File: source/openpower/keys.rst [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+-------------------------------------
+
+As explained above, using SSH Key Pairs to access an instance after a rebuild will *only* work with the **Original Key Pair**.
+
+So... What do you do if you lost the original key?
+
+Accessing an Instance After a Rebuild ... *Without the Original Key*
+--------------------------------------------------------------------
+
+As of right now, the preferred method (still being researched) is administrator password injection.
```
Comment (knightsamar) [inline source/openpower/keys.rst]: This is a potentially difficult method for us to implement on production because of the security risks posed by users setting a weak password and also since we have eliminated passwords completely for any kind of access to the instance.
Also, just to reset someone's password, I don't feel comfortable changing the configuration for the whole cluster.

Hence, I think this is good enough to know/document for us but let's make it clear are not going to support it on our cluster.

@ramereth thoughts?

File: source/openpower/keys.rst [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      ~$ cat ~/.ssh/<our_key>.pub | ssh <image_name>@<instance_IP_address> "mkdir -p ~/.ssh && cat > ~/.ssh/authorized_keys"
+      
+    The difference lies in the ``cat`` command, exchanging ``>>`` for ``>``
+    
+Changing a Key From OpenStack Web GUI
+-------------------------------------
+
+As far as in our knowledge, there is currently no way to edit the **Original Key Pair** designated to the instance.
+
+Not even through deleting the original key pair, creating a new one with the same name, and rebuilding the instance.
```
Comment (knightsamar) [inline source/openpower/keys.rst]: Please rephrase this to make it clearer under what condition it works/does not.

Also "same name" of what ?

File: source/openpower/keys.rst [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  
+      ~$ ssh <image_name>@<instance_IP_address>
+  
+  *Example:* ``~$ ssh centos@140.211.190.20``
+    
+  After we have tested that, we can ``Ctrl+D`` or ``logout`` out of the instance, and use ``cat`` and ``ssh`` to transfer ``<our_key>.pub`` to ``authorized_keys``, allowing multiple key pairs.
+  
+  - To **add** a key pair, retaining the original(s)::
+  
+      ~$ cat ~/.ssh/<our_key>.pub | ssh <image_name>@<instance_IP_address> "mkdir -p ~/.ssh && cat >> ~/.ssh/authorized_keys"
```
Comment (knightsamar) [inline source/openpower/keys.rst]: I would suggest `ssh-copy-id` here as @doublej472 did in his review.

File: source/openpower/keys.rst [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+Our best guess to the reason why is because when the instance is created, the **Original Key Pair** and its fingerprint are entered into the ``nova`` database, and there remain, uneditable, until its destruction.
+
+Note that we can use this knowledge to our advantage, however, as there is a fail-proof way to return the ``authorized_keys`` file to its original state by simply rebuilding the instance.
+
+Accessing an Instance After a Rebuild
+-------------------------------------
+
+As explained above, using SSH Key Pairs to access an instance after a rebuild will *only* work with the **Original Key Pair**.
+
```
Comment (knightsamar) [inline source/openpower/keys.rst]: What if I import a new key pair & rebuild the instance with another OS? 
Also, we should mention that if an instance is rebuilt, to access the instance we will have to use a **different** username.

File: source/openpower/keys.rst [hunk 6]
```diff
@@ -0,0 +1,106 @@
+.. _keys:
+
+SSH Key Behavior on OpenPOWER
+=============================
+
+These are findings of how SSH Keys behave in various conditions.
```
Comment (knightsamar) [inline source/openpower/keys.rst]: "...various conditions on OSL's POWERDev OpenStack cluster and how to manage keys"

File: source/openpower/keys.rst [hunk 7]
```diff
@@ -0,0 +1,106 @@
+.. _keys:
+
+SSH Key Behavior on OpenPOWER
```
Comment (knightsamar) [inline source/openpower/keys.rst]: It's also about "managing the keys" so the title should be suitably modified :)

File: source/openpower/keys.rst (hunk 1 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: should I leave the part about replacing the old key or just go through the steps to use ssh-copy-id?

File: source/openpower/keys.rst (hunk 5 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: The statement above saying "using SSH Key Pairs to access an instance after a rebuild will _only_ work with the **Original Key Pair**" holds true for every possible scenario so far _as long as the VM is rebuilt with the same OS and Endian-ness_. Updated in the next commit.

File: source/openpower/keys.rst (hunk 3 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: Edited to be more friendly wording, easier to understand the conditions.

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst (hunk 2 above)
Comment (ramereth) [inline source/openpower/keys.rst]: I think this should be documented internally and we just deal with manual ssh key resets via ticket requests.

**Comment by knightsamar** [comment]
@hodgesto, have you observed any special behaviour of keys when an instance is rebuilt from ubuntu to ubuntu ? specifically 16.04 ? We have an update on #29356 regarding this.

**Comment by hodgesto** [comment]
@knightsamar I have not specifically noticed anything, but I will test ubuntu (16.04) more thoroughly.

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst (hunk 2 above)
Comment (knightsamar) [inline source/openpower/keys.rst]: I think we should test this method out completely on our staging env before we ask customers to reach out to us in case their keys aren't working. Because this seems to be the only way ssh key resets works(correct me if I am wrong, @hodgesto)

We should test on staging:

1. How that setting would be reliably applied.
2. What is it's impact on machine and the access to machines for which we do not wish to reset the password?
3. What is it's impact on other dashboard users/projects?

File: source/openpower/keys.rst [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  
+  - To **add** a key pair, retaining the original(s)::
+  
+      ~$ ssh-copy-id <path/to/our_key>.pub <image_name>@<instance_IP_address>
+      
+    *Example:* ``~$ ssh-copy-id ~/.ssh/id_rsa.pub ubuntu@140.311.133.05``
+    
+  - To **change** the key pair, removing all others::
+
+      ~$ cat <path/to/our_key>.pub | ssh <image_name>@<instance_IP_address> "mkdir -p ~/.ssh && cat > ~/.ssh/authorized_keys"
```
Comment (knightsamar) [inline source/openpower/keys.rst]: Are you sure this works, if there is already a key pair ? I think `mkdir` will return an error and short-circuit it because the directory would already exist.

File: source/openpower/keys.rst [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+Changing a Key From OpenStack Web GUI
+-------------------------------------
+
+As far as in our knowledge, there is currently no way to edit the **Original Key Pair** designated to the instance.
+
+Once an instance is created, the original keypair with the original name and fingerprint will always be attached to that machine.
+
+The instance will *still* not change even if an administrator deletes the **Original Key Pair** and creates a new key pair with the same name as the **Original**, and rebuilds the machine.
+
+Our best guess to the reason why is because when the instance is created, the **Original Key Pair** and its fingerprint are entered into the ``nova`` database, and there remain, uneditable, until its destruction.
```
Comment (knightsamar) [inline source/openpower/keys.rst]: Say if we have a keypair named 'X-key' associated with an already running instance.

If we delete that keypair 'X-key', create a new keypair with exactly the same name but with a different pubkey and then rebuild the instance, what happens?

File: source/openpower/keys.rst (hunk 2 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: I agree with all that. And yes, so far, it's the only reliable method that a week or so of research has found. Also yes, it is a large security risk *but* I don't believe it would be so huge that it could shut down an entire cluster, or even a machine, unless there was a coordinated attack on that cluster in the few minutes that password injection was enabled. After due testing and research I believe it would be a valuable thing to support on our systems despite the potential security threat. If there is a way to limit this threat to a small amount of time, a single machine, or both, that would potentially be quite valuable.

Comment (hodgesto) [inline source/openpower/keys.rst]: That being said, I think in order to minimize security threats, the customers should still open a ticket and we should be the only ones able to do such a thing.

File: source/openpower/keys.rst (hunk 8 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: Yes, it has been tested. mkdir **-p** is recursive and just skips over making the directory if it already exists.

File: source/openpower/keys.rst (hunk 9 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: The instance does not change. The original *key* not just a link to the key via a name will always be restored to the VM upon rebuilding.

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst (hunk 9 above)
Comment (knightsamar) [inline source/openpower/keys.rst]: That's interesting and completely different from what I believed. We could link to an official doc explaining the fingerprinting behaviour.

Comment (hodgesto) [inline source/openpower/keys.rst]: That would be quite nice, if there was one. 
So far I have yet to find it if it exists.

File: source/openpower/keys.rst (hunk 2 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: update on this, it doesn't seem to be working right now. I think I may have to change the option as specified in the docs and reboot the _hypervisor_. But to preoperly change the hypervisor over the reboot I would need to change whatever chef recipe creates the hypervisor, which I don't particularly feel like I should be doing. Thoughts on how I should proceed? @knightsamar @ramereth

File: source/openpower/keys.rst (hunk 9 above)
Comment (knightsamar) [inline source/openpower/keys.rst]: So how did you figure it out ?

File: source/openpower/keys.rst (hunk 2 above)
Comment (knightsamar) [inline source/openpower/keys.rst]: On the staging chef run won't happen unless we trigger it from outside.
On the production, it happens every 30 minutes.

File: source/openpower/keys.rst (hunk 9 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: Just through some research on how the instances are created, what people from openstack said, a couple docs that aren't about this but talk about how different variables are stored in the nova database upon creation.

File: source/openpower/keys.rst (hunk 2 above)
Comment (hodgesto) [inline source/openpower/keys.rst]: Okay, but it wouldn't happen on reboot?

File: source/openpower/keys.rst [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+SSH Key Behavior and Management on OpenPOWER
+============================================
+
+These are findings of how SSH Keys behave in various conditions on OSL's POWERDev OpenStack cluster, and how to manage these keys.
+
+.. note::
+
+  For beginners: a good tutorial on SSH keys and SSH in general `here`__.
+  
+  And a quicker, more advanced  and jargon-full tutorial `here`__.
```
Comment (irandms) [inline source/openpower/keys.rst]: Extra space here after 'advanced'

Comment (hodgesto) [inline source/openpower/keys.rst]: Whoops 👍

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  <image_name> changes after rebuilding an instance with a different OS. Ubuntu machines -> ``ubuntu@<IP>`` CentOS machines -> ``centos@<IP>``, Debian machines -> ``debian@<IP>``, Fedora machines -> ``fedora@<IP>``, etc...
+
+Doing so will prompt the server and client to establish a secure connection, then the server authenticates the requesting client.
+
+If you have the correct credentials (the desired SSH private key), then you will be granted access to the server.
+
+.. note::
+
+  This description is very brief and not detailed enough to fully eclipse how SSH keys and the Diffie-Hellman exchange works. A more descriptive overview is linked in the section `above.`__
```
Comment (doublej472) [inline source/openpower/keys.rst]: Instead of saying `fully eclipse`, it should probably just be `fully explain`. Eclipse doesn't make that much sense in this context.

File: source/openpower/keys.rst [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+  ~$ ssh-keygen -f "<path/to>/.ssh/known_hosts" -R <instance_IP_address>
+  
+*Example:* ``~$ ssh-keygen -f "/home/scobeavs/.ssh/known_hosts" -R 140.211.123.456``
+
+This issue occurs because the client (the machine you're attempting to log on to the server through), is also attempting to authenticate the server, to keep the user (you) from connecting to a potentially dangerous imposter. This command removes the server from your 'known_hosts' file, allowing you to connect to the server as if it were brand new to you.
+
+Changing a Key From Inside an Instance
+--------------------------------------
+
+When we create our OpenPOWER OpenStack instance, we set up a key pair to create the machine with.
```
Comment (doublej472) [inline source/openpower/keys.rst]: Is this key already set up or do you mean that we need to create one?

File: source/openpower/keys.rst [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+  - To **change** the key pair, removing all others::
+
+      ~$ cat <path/to/our_key>.pub | ssh <image_name>@<instance_IP_address> "mkdir -p ~/.ssh && cat > ~/.ssh/authorized_keys"
+    
+    *Example:* ``~$ cat ~/.ssh/id_rsa.pub | ssh debian@140.244.313.10 "mkdir...``    
+    
+Changing a Key From OpenStack Web GUI
+-------------------------------------
+
+As far as in our knowledge, there is currently no way to edit the **Original Key Pair** designated to the instance.
```
Comment (doublej472) [inline source/openpower/keys.rst]: The beginning of this sentence is wordy, It might look better shorter or even removed up to the comma.

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by hodgesto** [comment]
Rebased and fixed all lint issues.

**Comment by ramereth** [comment]
@hodgesto I'm pretty sure you didn't rebase.. I shouldn't see my changes in the diff in this PR. Can you please verify and push again?

**Comment by hodgesto** [comment]
@ramereth was the Makefile and doc8.ini, etc. not your changes? Those are in this PR. If not, let me know I'll try again.

**Comment by ramereth** [comment]
@hodgesto it is but something funky is going on with the history. When I try to rebase to master I get conflicts which I shouldn't get if this has been rebased. I think what might help is if you compressed your commits into one and then we can redo the rebase again.

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by hodgesto** [comment]
@ramereth I think that did it. Your files are there as if I had started this PR after you merged #32.

**Comment by ramereth** [comment]
Please fix the following issues doc8 finds:

``` console
$ doc8 source/openpower/keys.rst
Scanning...
Validating...
source/openpower/keys.rst:156: D000 Literal block expected; none found.
source/openpower/keys.rst:162: D000 Bullet list ends without a blank line; unexpected unindent.
source/openpower/keys.rst:168: D000 Unexpected indentation.
source/openpower/keys.rst:169: D000 Block quote ends without a blank line; unexpected unindent.
source/openpower/keys.rst:41: D002 Trailing whitespace
source/openpower/keys.rst:46: D002 Trailing whitespace
source/openpower/keys.rst:60: D002 Trailing whitespace
source/openpower/keys.rst:61: D002 Trailing whitespace
source/openpower/keys.rst:62: D002 Trailing whitespace
source/openpower/keys.rst:75: D002 Trailing whitespace
source/openpower/keys.rst:82: D002 Trailing whitespace
source/openpower/keys.rst:104: D002 Trailing whitespace
source/openpower/keys.rst:107: D002 Trailing whitespace
source/openpower/keys.rst:110: D002 Trailing whitespace
source/openpower/keys.rst:116: D002 Trailing whitespace
source/openpower/keys.rst:117: D002 Trailing whitespace
source/openpower/keys.rst:134: D002 Trailing whitespace
source/openpower/keys.rst:145: D002 Trailing whitespace
source/openpower/keys.rst:35: D001 Line too long
source/openpower/keys.rst:61: D001 Line too long
source/openpower/keys.rst:102: D001 Line too long
source/openpower/keys.rst:107: D001 Line too long
source/openpower/keys.rst:116: D001 Line too long
source/openpower/keys.rst:117: D001 Line too long
source/openpower/keys.rst:134: D001 Line too long
source/openpower/keys.rst:154: D001 Line too long
========
Total files scanned = 1
Total files ignored = 0
Total accumulated errors = 26
Detailed error counts:
    - doc8.checks.CheckCarriageReturn = 0
    - doc8.checks.CheckIndentationNoTab = 0
    - doc8.checks.CheckMaxLineLength = 8
    - doc8.checks.CheckNewlineEndOfFile = 0
    - doc8.checks.CheckTrailingWhitespace = 14
    - doc8.checks.CheckValidity = 4
```

**Comment by hodgesto** [comment]
@ramereth doc8 now gives no errors for keys.rst

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by hodgesto** [comment]
+1 fixed  them.

File: source/openpower/keys.rst [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+On a simple level, instances can be accessed using ``ssh``::
+
+  $ ssh <image_name>@<instance_IP_address>
+
+*Example:* ``$ ssh ubuntu@140.211.123.456``
+
+.. note::
+
+  <image_name> changes after rebuilding an instance with a different OS. Ubuntu machines -> ``ubuntu@<IP>``
+  CentOS machines -> ``centos@<IP>``, Debian machines -> ``debian@<IP>``, Fedora machines -> ``fedora@<IP>``, etc...
```
Comment (ramereth) [inline source/openpower/keys.rst]: I would make the list of usernames and actual list so that it's easier to read. I'm pretty sure you can do that inside of a note.

File: source/openpower/keys.rst [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+granted access to the server.
+
+.. note::
+
+  This description is very brief and not detailed enough to fully explain how SSH keys and the Diffie-Hellman exchange works.
+  A more descriptive overview is linked in the section `above.`__
+
+.. __: keys.html#ssh-key-behavior-and-management-on-openpower
+
+*However*, a common issue occurs **after rebuilding an instance**, regarding your local 'known_hosts' file,
```
Comment (ramereth) [inline source/openpower/keys.rst]: Make ``known_hosts`` tech text.

File: source/openpower/keys.rst (hunk 2 above)
Comment (ramereth) [inline source/openpower/keys.rst]: I'd remove this whole section from the public wiki and state they would need to send a support ticket to us to get it fixed. I would then move this doc into our internal documentation.

**Comment by hodgesto** [comment]
@ramereth should I open a separate PR for adding the rest of this documentation to the internal docs?

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst [hunk 16]
```diff
@@ -29,8 +29,21 @@ On a simple level, instances can be accessed using ``ssh``::
...
+  | Debian   | ``debian``   |
+  +----------+--------------+
+  | CentOS   | ``centos``   |
+  +----------+--------------+
+  | Ubuntu   | ``ubuntu``   |
+  +----------+--------------+
+  | Fedora   | ``fedora``   |
+  +----------+--------------+
+  | OpenSUSE | ``opensuse`` |
+  +----------+--------------+
```
Comment (ramereth) [inline source/openpower/keys.rst]: You might use an in-line csv-table instead of this method for creating a table. We have a few examples in the repo.

**Comment by ramereth** [comment]
@hodgesto yes, please make a separate PR in the docs repo for that

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org

File: source/openpower/keys.rst [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+  This description is very brief and not detailed enough to fully explain how SSH keys and the Diffie-Hellman exchange works.
+  A more descriptive overview is linked in the section `above.`__
+
+.. __: keys.html#ssh-key-behavior-and-management-on-openpower
+
+*However*, a common issue occurs **after rebuilding an instance**, regarding your local ``known_hosts`` file,
+which stores a list of IP addresses and their corresponding DSA, RSA, or ECDSA host keys.
+
+If you are recieving something similar to the error::
```
Comment (knightsamar) [inline source/openpower/keys.rst]: _receiving_ or 'seeing' will also do

Comment (hodgesto) [inline source/openpower/keys.rst]: whoops :P

**Comment by osuosl-manatee** [comment]
This PR has been automatically built and can be viewed here: http://wiki-31.staging.osuosl.org
